### PR TITLE
Restore generation unwrap state if generate() raises

### LINKED ISSUE #6659: Generation context does not restore state after errors
### Reproduction

`_unwrap_model_for_generation` temporarily disables gradient checkpointing and, for gathered DeepSpeed ZeRO-3 generation, removes optimizer hooks. The cleanup statements are placed after `yield` without `finally` blocks. If generation raises inside the context, Python rethrows the exception at `yield` and skips both cleanup paths.

The gradient-checkpointing leak can be reproduced without distributed hardware:

```python
import types
from unittest.mock import patch

from trl.models.utils import _unwrap_model_for_generation


model = types.SimpleNamespace(is_gradient_checkpointing=True)
model.gradient_checkpointing_disable = lambda: setattr(model, "is_gradient_checkpointing", False)
model.gradient_checkpointing_enable = lambda: setattr(model, "is_gradient_checkpointing", True)
accelerator = types.SimpleNamespace(unwrap_model=lambda value: value)

try:
    with patch("trl.distributed.DistributedBackend", return_value=types.SimpleNamespace(is_zero3=False)):
        with _unwrap_model_for_generation(model, accelerator):
            raise RuntimeError("generation failed")
except RuntimeError:
    pass

print(model.is_gradient_checkpointing)
```

Output on `main` at `38b4b810`:

```text
False
```

The same control flow skips `add_hooks(model)` when an exception is raised from the gathered ZeRO-3 branch after `remove_hooks(model)`.

Temporary generation state should be restored whether the body exits normally or raises. The original generation exception should continue to propagate.

### System Info

- Platform: macOS-26.5.1-arm64-arm-64bit
- Python version: 3.12.13
- TRL base commit: 38b4b810
- PyTorch version: 2.13.0
- accelerator(s): MPS
- Transformers version: 5.14.1
- Accelerate version: 1.14.0
- Datasets version: 5.0.1
- HF Hub version: 1.26.0
- DeepSpeed version: not installed (the ZeRO-3 lifecycle is reproduced with a lightweight module stub)

### Checklist

- [x] I have checked that my issue isn't already filed.
- [x] I have included my system information.
- [x] Any code provided is minimal, complete, and reproducible.
- [x] Any code provided is properly formatted in code blocks.
- [x] Any traceback provided is complete.

### PR DESCRIPTION
**Pull request by YeonwooSung**
# What does this PR do?

`_unwrap_model_for_generation` disables gradient checkpointing and, on ZeRO-3, removes DeepSpeed hooks, then restores them only after a successful `yield`. If `generate()` raises, checkpointing stays off and the hooks stay removed.

This wraps the yield in `try`/`finally`, matching `_override_model_generation_config` in the same file. ZeRO-3 `add_hooks` is also in an inner `finally` so it runs even when generation fails inside `GatheredParameters`.

Adds `test_restores_gradient_checkpointing_on_error`.

Fixes #6659

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.

---

> [!NOTE]
> **Low Risk**
> Narrow bugfix to generation context cleanup; behavior on success is unchanged and failures now restore training state instead of leaving the model misconfigured.
> 
> **Overview**
> Fixes a cleanup gap in **`_unwrap_model_for_generation`**: when `generate()` (or other code inside the context) raises, gradient checkpointing could stay disabled and, on DeepSpeed ZeRO-3 with gathered params, optimizer hooks might not be reattached.
> 
> The yield path is now wrapped in **`try`/`finally`** so gradient checkpointing is re-enabled on exit, mirroring **`_override_model_generation_config`**. For ZeRO-3 with **`GatheredParameters`**, **`add_hooks`** runs in an inner **`finally`** so hooks are restored even if generation fails mid-context.
> 
> Adds **`test_restores_gradient_checkpointing_on_error`** to assert checkpointing is turned back on after a **`RuntimeError`** inside the context.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0139109ca116d57344a77ebb6912c6387c56f327. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by DimensionSTP** [comment]
Hi @YeonwooSung, I have a related fix in #6827 for preserving gradient-checkpointing kwargs during GKD generation. I compared both changes on the same current main, and they appear complementary rather than interchangeable.

Your inner `finally` restores the DeepSpeed hooks if generation raises. #6827 restores the original checkpointing kwargs on normal and exceptional exits, but does not yet cover that hook cleanup. Conversely, the parameterless `gradient_checkpointing_enable()` in this PR does not preserve explicitly configured kwargs such as `use_reentrant=False`.

I checked the combination locally with a CPU mock matrix covering both unwrap entry points, checkpointing on/off, explicit/default kwargs, and ZeRO-3 gather on/off. Combining the kwargs-restoring context from #6827 with your hook-cleanup `finally` passed all 72 cases; the existing model-utils tests also passed. This is not a new distributed GPU validation.

Would it make sense to land this cleanup first, then keep #6827 focused on the kwargs plumbing and the GKD call sites? Happy to coordinate so neither fix gets lost.

**Comment by YeonwooSung** [comment]
Agreed — complementary scopes. This PR stays on the #6659 exception path: restore checkpointing after `generate()` raises, and reattach ZeRO-3 hooks in the inner `finally`.

Kwargs restoration (`use_reentrant`, GKD call sites) stays in #6827. If this lands first, rebase #6827 on top and keep the inner `add_hooks` `finally`; thread `gradient_checkpointing_kwargs` through `disable_gradient_checkpointing` as you already do.

**Comment by DimensionSTP** [comment]
Thanks @YeonwooSung, I agree with that split. I'll keep #6827 focused on preserving gradient-checkpointing kwargs through the unwrap context and the GKD student and teacher call sites. Your PR handles exception-safe checkpointing restoration and ZeRO-3 hook cleanup.

I checked the combined behavior in the 72-case CPU matrix noted above. If #7162 lands first, I'll update #6827 on top of it, retain your inner add_hooks finally, and rerun the focused and distributed checks. I'll link this coordination in #6827 so reviewers can see why the two fixes are complementary.
